From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Hiroshige Hayashizaki <hiroshige@chromium.org>
Date: Tue, 4 Aug 2026 22:03:19 -0700
Subject: Move PrefetchService check a bit

Praparation for https://crrev.com/c/7952998.

This shouldn't impact the behavior much, as the `PrefetchService` check
failure is rare, and it's probably correct to always reject serving
when `PrefetchService` is gone.

Bug: 532959369, 497491557
Change-Id: I23288ff4401feccfeefa9a1af8f61a0ba03e3820
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/7950630
Reviewed-by: Taiyo Mizuhashi <taiyo@chromium.org>
Commit-Queue: Hiroshige Hayashizaki <hiroshige@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1673897}

diff --git a/content/browser/preloading/prefetch/prefetch_url_loader_interceptor.cc b/content/browser/preloading/prefetch/prefetch_url_loader_interceptor.cc
index 638d9eee13cb12d4a25f3fded1d9ceca4c1758e6..257e7f57d6e270f922eaca4d8c7280cff3e647ce 100644
--- a/content/browser/preloading/prefetch/prefetch_url_loader_interceptor.cc
+++ b/content/browser/preloading/prefetch/prefetch_url_loader_interceptor.cc
@@ -134,6 +134,15 @@ void PrefetchURLLoaderInterceptor::MaybeCreateLoader(
     return;
   }
 
+  PrefetchService* prefetch_service =
+      PrefetchService::GetFromFrameTreeNodeId(frame_tree_node_id_);
+  if (!prefetch_service) {
+    redirect_serving_handle_ = PrefetchServingHandle();
+    TRACE_EVENT_END("loading");
+    std::move(loader_callback_).Run(std::nullopt);
+    return;
+  }
+
   if (redirect_serving_handle_ &&
       redirect_serving_handle_.DoesCurrentURLToServeMatch(
           tentative_resource_request.url)) {
@@ -188,7 +197,7 @@ void PrefetchURLLoaderInterceptor::MaybeCreateLoader(
 
   TRACE_EVENT_END("loading");
   GetPrefetch(
-      tentative_resource_request.url,
+      *prefetch_service, tentative_resource_request.url,
       base::BindOnce(&PrefetchURLLoaderInterceptor::OnGetPrefetchComplete,
                      weak_factory_.GetWeakPtr(), tentative_resource_request.url,
 
@@ -198,6 +207,7 @@ void PrefetchURLLoaderInterceptor::MaybeCreateLoader(
 }
 
 void PrefetchURLLoaderInterceptor::GetPrefetch(
+    PrefetchService& prefetch_service,
     const GURL& url,
     base::OnceCallback<void(PrefetchServingHandle)> get_prefetch_callback)
     const {
@@ -205,14 +215,6 @@ void PrefetchURLLoaderInterceptor::GetPrefetch(
                     perfetto::Flow::FromPointer(
                         const_cast<PrefetchURLLoaderInterceptor*>(this)));
 
-  PrefetchService* prefetch_service =
-      PrefetchService::GetFromFrameTreeNodeId(frame_tree_node_id_);
-  if (!prefetch_service) {
-    TRACE_EVENT_END("loading");
-    std::move(get_prefetch_callback).Run({});
-    return;
-  }
-
   if (!initiator_document_token_.has_value()) {
     // TODO(crbug.com/40288091): Currently PrefetchServingPageMetricsContainer
     // is created only when the navigation is renderer-initiated and its
@@ -220,12 +222,13 @@ void PrefetchURLLoaderInterceptor::GetPrefetch(
     CHECK(!serving_page_metrics_container_);
   }
 
+
   auto callback = base::BindOnce(&OnGotPrefetchToServe, frame_tree_node_id_,
                                  url, std::move(get_prefetch_callback));
   auto key = PrefetchKey(initiator_document_token_, url);
   TRACE_EVENT_END("loading");
   PrefetchMatchResolver::FindPrefetch(
-      frame_tree_node_id_, *prefetch_service, std::move(key),
+      frame_tree_node_id_, prefetch_service, std::move(key),
       expected_service_worker_state_, serving_page_metrics_container_,
       std::move(callback),
       perfetto::Flow::FromPointer(
diff --git a/content/browser/preloading/prefetch/prefetch_url_loader_interceptor.h b/content/browser/preloading/prefetch/prefetch_url_loader_interceptor.h
index d3def650bf92a88cfd01501602840b7499307947..ca48558be05f9b1570a29a6654ab86f9845e94d4 100644
--- a/content/browser/preloading/prefetch/prefetch_url_loader_interceptor.h
+++ b/content/browser/preloading/prefetch/prefetch_url_loader_interceptor.h
@@ -18,6 +18,7 @@
 namespace content {
 
 class PrefetchContainer;
+class PrefetchService;
 class PrefetchServingPageMetricsContainer;
 class ServiceWorkerMainResourceHandle;
 
@@ -63,7 +64,8 @@ class CONTENT_EXPORT PrefetchURLLoaderInterceptor final
   // from `PrefetchService` and then goes through other checks in
   // `PrefetchUrlLoaderHelper`.
   // The |get_prefetch_callback| is called with this associated prefetch.
-  void GetPrefetch(const GURL& url,
+  void GetPrefetch(PrefetchService& prefetch_service,
+                   const GURL& url,
                    base::OnceCallback<void(PrefetchServingHandle)>
                        get_prefetch_callback) const;
 
